src/module/ovs/bpf.rs

```diff
@@ -0,0 +1,253 @@
+//! Rust<>BPF types definitions for the ovs module.
+//! Please keep this file in sync with its BPF counterpart in bpf/.
+
+use anyhow::{bail, Result};
+use plain::Plain;
+
+use crate::{
+    core::events::{
+        bpf::{parse_raw_section, BpfRawSection},
+        EventField,
+    },
+    event_field,
+};
+
+/// Types of events that can be generated by the ovs module.
+#[derive(Debug, Eq, Hash, PartialEq)]
+pub(crate) enum OvsEventType {
+    /// Upcall tracepoint.
+    Upcall = 0,
+    /// Upcall enqueue kretprobe.
+    UpcallEnqueue = 1,
+    /// Upcall return.
+    UpcallReturn = 2,
+    /// Upcall received in userspace.
+    RecvUpcall = 3,
+    /// Flow operation
+    Operation = 4,
+    /// Execute action tracepoint.
+    ActionExec = 5,
+    /// OUTPUT action specific data.
+    OutputAction = 6,
+}
+
+impl OvsEventType {
+    pub(super) fn from_u8(val: u8) -> Result<OvsEventType> {
+        use OvsEventType::*;
+        let owner = match val {
+            0 => Upcall,
+            1 => UpcallEnqueue,
+            2 => UpcallReturn,
+            3 => RecvUpcall,
+            4 => Operation,
+            5 => ActionExec,
+            6 => OutputAction,
+            x => bail!("Can't construct a OvsEventType from {}", x),
+        };
+        Ok(owner)
+    }
+}
+
+/// OVS Upcall data.
+#[derive(Default)]
+#[repr(C, packed)]
+struct UpcallEvent {
+    /// Upcall command. Holds OVS_PACKET_CMD:
+    ///   OVS_PACKET_CMD_UNSPEC   = 0
+    ///   OVS_PACKET_CMD_MISS     = 1
+    ///   OVS_PACKET_CMD_ACTION   = 2
+    ///   OVS_PACKET_CMD_EXECUTE  = 3
+    cmd: u8,
+    /// Upcall port.
+    port: u32,
+    /// Cpu ID
+    cpu: u32,
+}
+unsafe impl Plain for UpcallEvent {}
+
+pub(super) fn unmarshall_upcall(raw: &BpfRawSection, fields: &mut Vec<EventField>) -> Result<()> {
+    let event = parse_raw_section::<UpcallEvent>(raw)?;
+
+    fields.push(event_field!("upcall_port", event.port));
+    fields.push(event_field!("cmd", event.cmd));
+    fields.push(event_field!("cpu", event.cpu));
+    fields.push(event_field!("event_type", "upcall".to_string()));
+    Ok(())
+}
+
+/// OVS action event data.
+#[derive(Default)]
+#[repr(C, packed)]
+struct ActionEvent {
+    /// Action to be executed.
+    action: u8,
+    /// Recirculation id.
+    recirc_id: u32,
+    /// Queue id (only when tracking)
+    queue_id: u32,
```

Comment:
default known value is zero though I don't know if it could be even more confusing.

For postprocessing, I was thinking we should dump the configuration in the event file on some kind of header or something so that post-processors know what to expect. WDYT?

Reply:
> > default known value is zero though I don't know if it could be even more confusing.
> 
> I meant maybe using an i64 to store it and set -1 by default, not that I'm super convinced by this either. Looking at this again in this case can jhash return a valid hash of 0?
> 

Yes, that's why I said it could even be more confusing :-)

> > For postprocessing, I was thinking we should dump the configuration in the event file on some kind of header or something so that post-processors know what to expect. WDYT?
> 
> Not sure to follow the relation with the queue id. But in general yes, if I recall correctly we decided we could have events of all kinds, including configuration dumps. We just need to sync on that one to be on the same page, but IMHO the main missing piece is an helper to feed a custom event to the core.

This value is only generated if `--ovs-track` is provided. So if the post-processing engine knows what arguments were given to the module it knows `queue_id ` is irrelevant and should be ignored.

Reply:
Agree. The `queue_id` is only present on hooks that are loaded only when `--ovs-track` is set. All except this one. So I can move it to another event type and only send it if needed. It seems a bit wasteful to add a section header for only a 32 bit field but it's probably better than the confusion that could be generated otherwise.